### PR TITLE
Changes from travisdowns/libpfc

### PR DESCRIPTION
**Pull request by travisdowns**
These are a few accumulated changes that I've made to libpfc on my branch. Feel free to reject anything you don't like, but ideally I'd like to keep my branch in sync (or remove the need for it entirely).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,6 @@
 build
+kmod.build
```
Comment (travisdowns) [inline .gitignore:2]: Add a few other ignores to .gitignore (see the Makefile build below).

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,32 @@
+.PHONY : all clean
```
Comment (travisdowns) [inline Makefile:1]: I added a Makefile for libpfc, which is totally parallel to the meson/ninja stuff. My project is small and make-only, and it seems to be a pain to ask uses to install *two* other build tools just to build the libpfc part, so I added this to simplify the process for them. It pretty much does the same thing as ninja/meson, but outputs to a different location. If you want to keep this out of your branch that's fine, I'll try to cherry-pick the other commits in that case.

File: include/libpfc.h [hunk 3]
```diff
@@ -12,6 +12,13 @@
 typedef uint64_t  PFC_CFG;
 typedef int64_t   PFC_CNT;
 
+// the next three definitions are the positions of the three fixed counters within
```
Comment (travisdowns) [inline include/libpfc.h:15]: Added defns for the fixed counters which seem to be "architectural" so they are consistent.

File: include/libpfc.h [hunk 4]
```diff
@@ -121,6 +128,10 @@ _pfc_asm_code_(op)                              \
 
 void      pfcRemoveBias     (PFC_CNT* b, int64_t mul);
 
+/**
+ * Return a string representation of an error code such as those returned by pfcInit
+ */
+const char *pfcErrorString(int err);
```
Comment (travisdowns) [inline include/libpfc.h:134]: In several places I added some error handling, and some methods now return a specific error code. You can pass the code in there to get an error string, which I use for display. If the error code isn't known (e.g., for some methods that still return 1), you just get "Unknown error".

File: kmod/pfckmod.c [hunk 5]
```diff
@@ -56,6 +58,16 @@ static ssize_t pfcCntWr(struct file*          f,
                         char*                 buf,
                         loff_t                off,
                         size_t                len);
+static ssize_t pfcVerboseRd(struct kobject* kobj,
```
Comment (travisdowns) [inline kmod/pfckmod.c:61]: These two methods, along with the PRINTV above, allow you to enable a "verbose mode" in the kernel driver (by writing `1` to `/sys/modules/pfc/verbose`, which will spit to the kernel log a bit more info useful for diagnosis (I haven't actually used it much).

File: kmod/pfckmod.c [hunk 6]
```diff
@@ -56,6 +58,16 @@ static ssize_t pfcCntWr(struct file*          f,
...
+						struct kobj_attribute* attr,
+						char* buf);
+static ssize_t pfcVerboseWr(struct kobject* kobj,
+						struct kobj_attribute* attr,
+						const char* buf,
+						size_t count);
+static ssize_t pfcCR4PceRd(struct kobject* kobj,
```
Comment (travisdowns) [inline kmod/pfckmod.c:68]: This is a new method which exposes the value of CR4.PCE, so userland can check it and fail gracefully if it is zero.

File: kmod/pfckmod.c [hunk 7]
```diff
@@ -100,19 +114,43 @@ static const struct bin_attribute   PFC_ATTR_counts     = {
 	.read    = pfcCntRd,
 	.write   = pfcCntWr
 };
-static const struct bin_attribute*  PFC_ATTR_GRP_LIST[] = {
+static const struct bin_attribute*  PFC_BIN_ATTR_GRP_LIST[] = {
```
Comment (travisdowns) [inline kmod/pfckmod.c:117]: Changed to BIN_ATTR group, contrasting with the new STR_ATTR group which has the verbose and cr4.pce flags.

File: src/libpfc.c [hunk 8]
```diff
@@ -382,7 +382,28 @@ static const struct EVENT EVENT_LIST[256] = {
     {0x00, NULL             , NULL}
 };
 
+/* Error Codes */
+#define PFC_ERR_OPENING_SYSFILE (-10)
```
Comment (travisdowns) [inline src/libpfc.c:386]: Here are the error codes, but I should probably move them to the libpfc.h header file so clients could check them directly?

File: src/libpfc.c [hunk 9]
```diff
@@ -382,7 +382,28 @@ static const struct EVENT EVENT_LIST[256] = {
...
+#define PFC_ERR_CR4_PCE_NOT_SET (-14) // drive reported that cr4.pce wasn't set, or there was somehow an issue reading it
+#define PFC_ERR_AFFINITY_FAILED (-15) // setting CPU affinity failed (perhaps affinity is set externally excluding CPU 0?)
+#define PFC_ERR_READING_MASKS   (-16) // didn't read the expected number of mask bytes from the sysfs
 
+typedef struct {
+    int errorCode;
+    const char *message;
+} ErrorInfo;
+
+ErrorInfo AllErrors[] = {
```
Comment (travisdowns) [inline src/libpfc.c:399]: The table we use for lookup of the error strings. So to add a new error you just add a `#define` above and a line in this table.

File: src/pfcdemo.c [hunk 10]
```diff
@@ -64,6 +65,34 @@ static const int         NUMCOUNTS;
 int main(int argc, char* argv[]){
 	int i;
 	
+	int verbose = 0, dump = 0, option;
```
Comment (travisdowns) [inline src/pfcdemo.c:68]: Make `pfcdemo` a bit richer, with a couple of command line options, for dumping and enabling verbose mode.

File: src/pfcdemo.c [hunk 11]
```diff
@@ -64,6 +65,34 @@ static const int         NUMCOUNTS;
...
+			break;
+		default:
+			fprintf(stderr, "Usage: pfcdemo [-d] [-v]\n"
+					"\t-d\n\t\tDump the list of available events and quit\n"
+					"\t-v\n\t\tVerbose output\n");
+			exit(1);
+		}
+	}
+
+	(void)verbose;
```
Comment (travisdowns) [inline src/pfcdemo.c:89]: Actually we don't even do anything special for verbose yet, huh.

**Comment by obilaniu** [comment]
@travisdowns Thanks. Eventually I'll have to put some thought into organizing the source code more seriously, writing in proper kernel style, and so on and so forth.

I much prefer Meson because it's simple and I can get away from the autotools hell, but admittedly this project currently amounts only to a couple source files and so it's probably (still) tolerable. But Meson is on the up; Projects as large as GNOME, X.org and Mesa are using it.

I think the Makefile you committed leaks a bit of info about your personal filesystem hierarchy?

---

I've also been thinking about adding support for making counter values migrate with the process using `preempt_notifier_register()` in the kernel, but I have no easy way of checking that a given `struct task_struct*` is or is not using counters, and what were their counts as of the last time they were preempted. Thoughts?

**Comment by travisdowns** [comment]
Oops, you are right about the path (I had copy-pasted it from the meson generated stuff), will remove it in next commit.

I'm not a fan of autotools by any means and I'm not using it here. Just a plain makefile is fine for what boils down to the 2 source files. The kernel build you don't have another choice so I'm not counting that. 

I don't expect you to keep the Makefile up to date or anything and I can add a disclaimer that it's not the default build method. Now meson may be great, but I would say more 50% of my time trying to get `libpfc` has been working around a couple of different meson issues, so perhaps that's what frustrated me. Keep in mind there are very different priorities for a huge project with millions of LOC and very long builds compared to very small open-source projects. The former could prefer more obscure build methods if they give them an important advantage in compile time, determinism, error handling, etc - but for smaller projects they probably just want "simplest setup"?

I'm not quite sure on how I'll consume `libpfc` long term - right now it's just a git submodule so people have to build it, but if there are release binaries I could just include those (ofc the `pfc.ko` always has to be built to align with the users' local kernel), rather than forcing users to build everything.

**Comment by travisdowns** [comment]
About `preempt_notifier_register()` I guess it depends on how much of `perf_events` you want to rebuild. I feel like if context switches are likely, the counts are already getting so large than the fast userspace read and bias removal features of `libpfc` are much less important, and those are the two main reasons to use it over simply using `perf_events`. 

That said, if you want to restore counters across context switches, I could check out how `perf_events` does it, perhaps just for ideas, but perhaps you can in fact just use the existing support. I know `perf_events` saves the counter values when the process is context switched _out_, but I don't know if it simply restores them when it is switched back _in_, or if it just tracks the accumulated values and sets them all to zero each quantum. I suspect more the latter.

For my particular purpose (benchmarking), what I plan to do is just to set one counter to count HW_INTERRUPTS and if non-zero just retry. Other users may not find that feasible, however!

**Comment by obilaniu** [comment]
@travisdowns 

> or if it just tracks the accumulated values and sets them all to zero each quantum. I suspect more the latter

Ow, that might be troublesome. If the code gets preempted between `PFC_START()/PFC_END()`, you would get a meaningless value when you make the userland `rdpmc` reads (!), which were the whole point of the kernel module. I guess the solution would really be to document that you can expect overflow sooner than you might expect (because the counters are 32 or 40 or 48 or whatever bits wide).

In my mind I've also floated the idea of a per-monitored-thread, vDSO-like 4KB page that would contain a preemption count, the counter values, configs and masks as of last preempt and maybe even the appropriate `PFC_START/PFC_END` code snippets for the current counter config. But I've also got few ideas about how to implement such a thing.

**Comment by travisdowns** [comment]
@obilaniu - what you describe in the last paragraph is almost _exactly_ how user-mode `perf_events` `rdpmc` works. You `mmap` a page which has the accumulated counter values, configuration information, etc. Then to do a read, you read from the mmaped page, do your `rdpmc` and then read again the `mmaped` page. You add the values you get from the `rdpmc` to the accumulated values in the page (the kernel keeps track of updating them on context switches and overflow). This also takes care of a context switch that happens in the critical region (since the second read will see a different `lock` value, which means try again).

I really recommend checking out the stuff under "MMAP layout" and "cap_user_rdpmc" [here](http://man7.org/linux/man-pages/man2/perf_event_open.2.html). It will be a lot of work to reinvent the wheel.

**Pull request merged**